qbe_backend/src/lowering_context.rs

```diff
@@ -0,0 +1,48 @@
+use std::collections::BTreeMap;
+
+use crate::lower_lang::*;
+
+pub struct LoweringCtx {
+    prefix: String,
+    pool: BTreeMap<String, String>,
+    uniq: u32,
+}
+
+impl LoweringCtx {
+    pub fn new(prefix: &str) -> LoweringCtx {
+        LoweringCtx {
+            prefix: prefix.to_string(),
+            pool: Default::default(),
+            uniq: 0,
+        }
+    }
+
+    pub fn uniq_name(&mut self, prefix: &str) -> String {
+        let name = format!("{}{}", prefix, self.uniq);
+        self.uniq += 1;
+        name
+    }
+
+    pub fn name_for_str(&mut self, str: &str) -> String {
+        let len = self.pool.len();
+        self.pool
+            .entry(str.to_string())
+            .or_insert_with(|| format!("{}_{}", self.prefix, len))
+            .to_string()
```

Comment:
This is a good catch but currently the pool is append only. Trying to use `uniq_name` makes the borrow checker angry:

```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
  --> qbe_backend/src/lowering_context.rs:30:32
   |
30 |             .or_insert_with(|| self.uniq_name(&self.prefix))
   |                                ^^^^^---------^------------^
   |                                |    |         |
   |                                |    |         immutable borrow occurs here
   |                                |    immutable borrow later used by call
   |                                mutable borrow occurs here

error[E0500]: closure requires unique access to `*self` but it is already borrowed
  --> qbe_backend/src/lowering_context.rs:30:29
   |
28 |         self.pool
   |         --------- borrow occurs here
29 |             .entry(str.to_string())
30 |             .or_insert_with(|| self.uniq_name(&self.prefix))
   |              -------------- ^^ ---- second borrow occurs due to use of `*self` in closure
   |              |              |
   |              |              closure construction occurs here
   |              first borrow later used by call

Some errors have detailed explanations: E0500, E0502.
For more information about an error, try `rustc --explain E0500`.

```